### PR TITLE
fmt v11 and c++20 modules support

### PR DESCRIPTION
**Pull request by kamrann**
Updates to address issues in modules support provided by the build2 package. See issue #5.

### CODE REVIEW

**Comment by kamrann** [comment]
Notes and various issues encountered so far:

1. It appears the upstream modularization of `fmt` is currently making assumptions of more than just c++20, since it doesn't wrap includes such as `<expected>` in preprocessor conditionals within the module source file, as it does with the same includes in the header mode. This leads to a warning on MSVC unless using both `/std:latest` (`cxx.std=experimental`) and latest preview 14.41.
2. MSVC will error out when building `fmt` module unless `cc.reprocess = true` is set. This looks like it's probably a bug in MSVC when combining `/P` with a private module fragment. Unfortunately `cc.reprocess = true` combined with modules appears to be somewhat bugged in `build2` generally, clean builds are ok but iterative builds keep generating errors about missing preprocessed source files.
3. clang with libc++ on Windows looks like it just doesn't work with `fmt`, modules or no, some issue relating to interpretation of compiler defines and `FMT_MSC_VERSION `. Needs more investigation, though not a priority of course.
4. clang with MSVC on Windows also giving errors about internal linkage, didn't dig further.
5. The `fmt` module source file is in `/src`. Since it's an interface unit though, I think this means we'll need to tweak the `install` behaviour in the buildfile to install this source file. Obviously all the interface stuff is in the header, but since BMIs aren't installed, I'd assume downstream consuming build tools working with the installation would need access to the source file in order to rebuild it?

**Comment by kamrann** [comment]
Re. 2
build2 behaviour with `reprocess=true`: https://github.com/build2/build2/issues/409
MSVC `/P` and private module fragments: https://developercommunity.visualstudio.com/t/Separate-preprocessing-with-P-fails-wit/10707183

**Comment by kamrann** [comment]
Some more experimenting. 4. also applies with clang on Linux. I'm unclear on where the bugs are, but I've filed an issue with both [LLVM](https://github.com/llvm/llvm-project/issues/99825) and [fmt](https://github.com/fmtlib/fmt/issues/4081).

File: fmt-tests/basics/buildfile [hunk 1]
```diff
@@ -0,0 +1,45 @@
+include ../test-main/
+
+#test_names =
```
Comment (Klaim) [inline fmt-tests/basics/buildfile]: Are these comments still useful?

File: fmt-tests/basics/buildfile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#  args-test \
+#  assert-test \
+#  base-test
+gtest_test_names = args-test assert-test base-test chrono-test color-test compile-fp-test compile-test format-test gtest-extra-test noexception-test os-test ostream-test printf-test std-test unicode-test xchar-test
```
Comment (Klaim) [inline fmt-tests/basics/buildfile]: Could this list have been formed from globing the `.cc` files? If not consider adding a comment about where to look to update this list (in `upstream/tests/CMakeFiles.txt`)

File: fmt-tests/basics/buildfile [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+# header-only-test <- header only mode seemingly not supported by this build2 package?
```
Comment (Klaim) [inline fmt-tests/basics/buildfile]: Yes it was not supported, I basically had the same reasons as explained [in the packaging guide](https://build2.org/build2-toolchain/doc/build2-toolchain-packaging.xhtml#dont-header-only) although that wouldnt prevent support of header-only mode later if the demand is there. So far nobody needed it.

File: fmt-tests/test-main/buildfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+
+libs =
+import libs += fmt%lib{fmt}
+import libs += gtest%lib{gtest} gmock%lib{gmock}
+
+liba{test-main}: test/cxx{test-main gtest-extra util} test/hxx{gtest-extra mock-allocator test-assert util} $libs
```
Comment (Klaim) [inline fmt-tests/test-main/buildfile:6]: Would defining it as `libu` would have helped usage?

File: fmt/tests/basics/buildfile [hunk 5]
```diff
@@ -1,9 +1,13 @@
 import! [metadata, rule_hint=cxx.link] libs = fmt%lib{fmt}
 
-./ : exe{driver} : {cxx}{driver} hxx{tests.inl} $libs testscript{**}
+./ :
```
Comment (Klaim) [inline fmt/tests/basics/buildfile]: Is that  line still necessary?

**Review by Klaim — APPROVED** [review]
LGTM overall, most of my comments are for clarifications

File: fmt/tests/basics/buildfile (hunk 5 above)
Comment (kamrann) [reply fmt/tests/basics/buildfile]: Apparently not ;)
I may have thought it more clear since the following two are both conditional, but indeed seems superfluous.

File: fmt-tests/test-main/buildfile (hunk 4 above)
Comment (kamrann) [reply fmt-tests/test-main/buildfile:6]: Honestly, I'm yet to understand the intended usage of `libu*`, despite having read through the docs information on them many times. When/why do you choose to use it?

I also found this: ```Note also that if you plan to link any of your unit tests in the whole archive mode, then you will also need to exclude the source file containing the primary executable's main() from the utility library```
which suggests that if used here then `test-main` library could not contain `text-main.cc`. Having said that, the `test-main` library target is essentially reflecting upstream CMake and it's arguably not even worth having?
